### PR TITLE
SI-5722 Macrofy classOf, make it work with ClassTags

### PR DESCRIPTION
**Pull request by soc**

### CODE REVIEW

**Comment by retronym** [comment]
We'll have to address the unaddressed review comments from Paul's PR #1642.

This one should be split into to commits; one to macro-ify classOf, and a second to extend it to support class tags, with suitable pos and neg tests and commit messages.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 1]
```diff
@@ -3822,47 +3822,37 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
...
-        typedTypeApply(tree, mode, fun, args1)
+        if (sameLength(args, tparams)) {
+          //@M! the polytype denotes the expected kind
+          val args1 = map2Conserve(args, tparams)((arg, tparam) => typedHigherKindedType(arg, mode, GenPolyType(tparam.typeParams, AnyClass.tpe)))
+          typedTypeApply(tree, mode, fun, args1)
+        } else TypedApplyWrongNumberOfTpeParametersError(fun, fun)
+        // @M: there's probably something wrong when args.length != tparams.length... (triggered by bug #320)
+        // Martin, I'm using fake trees, because, if you use args or arg.map(typedType),
+        // inferPolyAlternatives loops...  -- I have no idea why :-(
+        // ...actually this was looping anyway, see bug #278.
```
Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:3833]: Not sure whether the new placement of the comment is clear.

**Comment by retronym** [comment]
I'm going to close this for now as I'm trying to trim our PR queue down to size. Please resubmit when this is ready for review.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 2]
```diff
@@ -3822,47 +3822,37 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
...
       case SingleType(_, _) =>
         typedTypeApply(tree, mode, fun setType fun.tpe.widen, args)
-      case PolyType(tparams, restpe) if tparams.nonEmpty =>
+      case PolyType(tparams, restpe) =>
```
Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:3836]: Why was the guard removed here?

**Comment by soc** [comment]
@retronym Agree.

**Pull request closed without merging**